### DISCUSSION TITLE
Plan to use in car. CAN-BUS

### DISCUSSION THREAD

**Discussion by kebibg** [HASwitchPlate/openHASP#442] [Ideas]
Hi,
After adding an offline mode will be good
to have CAN-BUS (TWAI) support. Then we can 
use it in a car. Can make an additional dashboard.
Or use it in industrial components.

Regards

**Comment by fvanroie** [comment]
I don't have any knowledge or experience with CAN-BUS networks. To my understanding the messages are only a few bytes in length.
I'm curious to know how do you envision the communication in what openHASP using Can-Bus?

**Reply by kebibg** [reply]
fvanroie,
Thanks for your answer. As I write above I need to use it in a car or truck as an additional gauge.
I am familiar with communication but with user interfaces not. OpenHASP is a very good and intuitive user interface.
At this time with their (openHASP) functions, I can make this communication with another one esp32 with can-bus(TWAI) and MQTT
server in it, then a wireless connection between 2 modules. This method will be more flexible but more expensive.
In short, the answer to your question is that I want to use your product for a different purpose.

Regards

**Reply by fvanroie** [reply]
Thank you for clarifying your use-case.

I am mostly interested in the low level implementation of the Canbus to MQTT protocol. Is that firmware available somewhere?
You are only sending values to the GUI to update the gauge. So there is no two-way communication from the UI to to the Car/Truck?

**Reply by kebibg** [reply]
Hi, fvanroie
Sorry for the late answer. If you want I can give you my code for reading DPF from Opel Corsa D.
ESP 32 has a CAN-BUS controller into it. You need just a driver IC to use it. Have 2 types of messages in this communication.
Short and long. Have 2 types of IDs 11bit and 29bit. Data always is 8 bytes. For long messages data is sent on a potion of 8 bytes.

Regards

**Reply by marcojk** [reply]
Hi @kebibg,
quite OT here, but can you please report the frame/id/payload to read the DPF on a Corsa? I suppose it's on the 500kbs bus...
Thanks

**Reply by kebibg** [reply]
Hi @marcojk ,
You need to use mode AA!! This is GM-enhanced diagnostic mode. The PIDs are 0129 and 012A. You need to read 8 bytes.
In the next example, I read from CAN-BUS 500K with filter 5E8(because enhanced diagnostic UUDT answer ID is different from standard 7E8!):
AA - ReadDataByPacketIdentifier
0129 and 012A - PID who need to read
 
AA0129:  5d 13 05 00 0c 00 00
AA012A:  00 1a 5c 00 03 31 00

You need to send it to address 7E0, length 03.  AA0129 or AA012A. 

0129 REGENS AFTER DPF EXCHANGE, DPF LOAD STATE, DPF DIFFERENTIAL PRESSURE(KpA), DPF DIFFERENTIAL PRESSURE(VOLT), 
A:5D
B:13
C:05
D:00
E:0C
F:00
G:00

A - Soot A*200/255 in %
B - DIFFERENTIAL pressure sensor in Voltage B*5.5/255
C - DIFFERENTIAL pressure sensor in kPa
D
E - REGENERATIONS AFTER DPF EXCHANGE d(hex)= 12

012A dISTANCE FROM REPLACEMENT, DISTANCE FROM LAST REGENERATION, REGENERATION STATE 00-OFF 255-100%
A:00 
B:1A
C:5C
D:00
E:03
F:31
G:00

A+B+C dISTANCE FROM REPLACEMENT
D+E+F DISTANCE FROM LAST REGENERATION
G   REGENERATION STATE 00-OFF 255-100%

this is the working code for ESP32 and Oled lcd:

[https://drive.google.com/drive/folders/1DZZOMxNTfSzW7udh2gvtrMUesNHjIoNK?usp=sharing](url)

Regards

P.S. This is 1.3CDTI engine from Fiat!!!

If you are interested I can give you a document from GM with all specifications in ENHANCED diagnostic just 330 pages.

**Reply by marcojk** [reply]
Thanks kebibg, great deal of information.
I used a request like this, which is another service
    msg.data[0] = 0x03;
    msg.data[1] = 0xaa;
    msg.data[2] = 0x03;
    msg.data[3] = 0xFA;
Unfortunately this worked with engine Z13DTC (the same FIAT+GM as yours) but gave garbage data with engine A13DTC.
I'll test your code, and check the document if you can provide it ;) thanks a lot!

**Reply by Tomi8609** [reply]
I'd be very interested in the 330-page GM documentation. 
Thank you in advance.

**Reply by kebibg** [reply]
Hi,
This and 2 additional files are in my drive.
https://drive.google.com/open?id=1rZfA1cYeLvnYF3p9sTm3SzxWFSyUFN4o&usp=drive_fs

**Comment by matsekberg** [comment]
I also have ides to use this great user interface solution in my RV. But I think the solution is to use another ESP32 to communicate with the vehicle since there is hardly a on-to-one mapping of values on the bus and values to display. Also, CANBUS is not really in the same scope as a generic user interface component. I would recommend a Arduino software on the ESP32 or try to use EspHome as a interface to the vehicle and communicate with openHASP over a serial port.

**Comment by kebibg** [comment]
Hi,
My code is for Z13DTJ.

Regards

**Comment by Tomi8609** [comment]
Hi!
Thank you very much

kebibg ***@***.***> ezt írta (időpont: 2026. ápr. 17., P,
5:41):

**Reply by kebibg** [reply]
You are welcome.

**Comment by Tomi8609** [comment]
I have a question. I looked through the documentation you sent, but
unfortunately I couldn't find what I was looking for.
I’m building a CAN bus decoder for my Opel Zafira B 1.9 CDTi. I’ve already
been able to decode a lot of data from the MS and LS buses; now only the
HS-CAN remains. Currently, I’m using the same diagnostic query as the
OP-COM software. The ECU responds and sends the following:
ID: 0x5E8 DLC: 8 DATA: 03 00 00 08 00 00 80 00 STD
ID: 0x5E8 DLC: 8 DATA: 04 37 93 04 00 05 00 00 STD
ID: 0x5E8 DLC: 8 DATA: 10 8D 0C A4 00 20 80 00 STD
ID: 0x5E8 DLC: 8 DATA: 11 4E 76 6C 58 3A 0C 06
ID: 0x5E8 DLC: 7 DATA: 12 00 01 67 77 01 5F
ID: 0x5E8 DLC: 8 DATA: 13 31 18 00 00 00 0C 00
ID: 0x5E8 DLC: 5 DATA: 14 50 00 4C 0F
ID: 0x5E8 DLC: 8 DATA: 15 FF 00 0A 4C C4 00 7F
ID: 0x5E8 DLC: 7 DATA: 16 00 7D 00 00 00 00
ID: 0x5E8 DLC: 8 DATA: 17 04 16 0B 35 25 32 1F
ID: 0x5E8 DLC: 7 DATA: 18 2B 49 37 00 29 00
ID: 0x5E8 DLC: 8 DATA: 19 44 98 82 00 12 32 10
ID: 0x5E8 DLC: 8 DATA: 1A 00 00 00 67 BD 00 00
ID: 0x5E8 DLC: 8 DATA: 1B 02 1C 99 85 03 E2 1D STD
ID: 0x5E8 DLC: 6 DATA: 1C 38 0A BB 0A 8C
ID: 0x5E8 DLC: 6 DATA: 1D 00 00 00 5C 58
ID: 0x5E8 DLC: 8 DATA: 1E 00 00 C7 80 2D 80 00
ID: 0x5E8 DLC: 3 DATA: 1F 00 00
ID: 0x5E8 DLC: 5 DATA: 21 00 00 00 00
ID: 0x5E8 DLC: 5 DATA: 22 0C 00 19 00

The ECU updates and continuously sends this data.
I’ve already decoded quite a lot from this data:
Battery voltage, speed, RPM, coolant temperature, outside temperature,
charge air temperature, DPF saturation %, distance traveled since DPF
regeneration, total mileage. What’s still missing is turbo pressure, rail
pressure, air pressure, throttle valve %, pedal %, EGR %...
Do you happen to have any decoding formulas for these?

Best regards,
Tamás

kebibg ***@***.***> ezt írta (időpont: 2026. ápr. 17., P,
7:47):

**Reply by kebibg** [reply]
Hi,
I stop play with this. I sell my Corsa and now I am owner of Insignia A, Insignia B and Astra K. All of them are diesel. If torque soft decide this parameters you can see their calculation. Tomorrow I can send you additional PIDs for GM torque with their calculations.

**Comment by Tomi8609** [comment]
Ok, thanks, i accept.

kebibg ***@***.***> ezt írta (időpont: 2026. ápr. 17., P
15:47):
